In [26]:
import pandas as pd

data = pd.read_csv('C:/Users/Rob/Desktop/Kaggle Projects/data-exercise/data.csv')
data.head()

trip_id driver         lon       lat                 date   speed  \
0        1   Alan -122.128910  37.44529  2009-04-05 13:39:54  11.250   
1        1   Alan -122.125960  37.44332  2009-04-05 13:40:24  18.750   
2        1   Alan -122.128170  37.44071  2009-04-05 13:40:54  38.125   
3        1   Alan -122.131740  37.43711  2009-04-05 13:41:24  36.250   
4        1   Alan -122.132805  37.43598  2009-04-05 13:41:54   0.000   

   direction  odometer event  
0        185    743984     N  
1        117    744344     Y  
2        218    744713     N  
3        218    745222     N  
4        218    745380     N


# Q1 

First thing to do is split the data by driver, putting the respective driver into their own dataframe

In [27]:
alan = data[data['driver'] == 'Alan']
barbara = data[data['driver'] == 'Barbara']

Next we can group each dataframe and get the number of times each event occurred in each dataframe

In [37]:
print('Alan: ',alan.groupby('event').size())
print('')
print('Barbara: ',barbara.groupby('event').size())

Alan:  event
N    183
Y     40
dtype: int64

Barbara:  event
N    610
Y     15
dtype: int64


Even though Alan has more 'Y' occurrences, he may be the safer driver due to more driving, hence more opportunities for events.
To check whether this is, or isnt the case, we normalize the data.  To normalize the data, I take the count of the occurences, and divide it by the total number of occurrences

In [49]:
alan_n = alan.groupby('event').size()[0]/(alan.groupby('event').size()).sum()

barbara_n = barbara.groupby('event').size()[0]/(barbara.groupby('event').size()).sum()

print('Percent of times Alan had an \'event\': %.2f' % (1 - alan_n))
print('Percent of times Barbara had an \'event\': %.2f' % (1 - barbara_n))

Percent of times Alan had an 'event': 0.18
Percent of times Barbara had an 'event': 0.02


**Question 1 Answer**:

Barbara had on average 0.02 events every 30 seconds, as opposed to Alan who had an event 0.18 times every 30 seconds.  Since we assume events as roughly correlated with driver safety we can conclude that Barbara is the safer driver since she had less events on average every 30 seconds.

Other thoughts, I don't believe any of the other columns provide any insight into 'safety' of a driver, speed maybe, but without knowing the legal speed limit of the driver at that time not much can be done.


# Question 2

First thing to do is split the data into trip 7/trip 13


In [51]:
trip_7 = data[data['trip_id'] == 7]
trip_13 = data[data['trip_id'] == 13]

Next, take a quick look at the dataframes, luckily they're small enough to be printed right in the window.

In [50]:
print(trip_7['speed'])
print(trip_13['speed'])

158    40.000
159    38.750
160    53.750
161    59.375
162    64.375
163    65.625
164    28.750
165     0.000
166     0.000
167     0.000
168     8.125
Name: speed, dtype: float64
248    18.750
249     9.375
250     8.125
251    19.375
252    13.125
253    18.750
254    15.000
255    10.625
256     0.000
257     0.000
258     0.000
259     6.250
260    16.250
261     0.000
262     0.000
263     0.000
264     5.625
265    14.375
266    20.000
267     0.000
268     0.000
269     6.250
270     0.000
271     0.000
272     0.000
273    16.875
274     0.000
275    18.125
276     0.000
277     0.000
278    18.750
279    29.375
280    26.250
281    18.125
282    15.625
283    25.000
284     8.125
285    19.375
286    15.000
287    31.250
288    30.625
289    16.875
290    16.250
Name: speed, dtype: float64


Now to perform a quick analysis of the average speed of each trip

In [23]:
import numpy as np

print("Trip 7 Average Speed: %.1f mph" % np.mean(trip_7['speed']))
print("Trip 13 Average Speed: %.1f mph" % np.mean(trip_13['speed']))

Trip 7 Average Speed: 32.6 mph
Trip 13 Average Speed: 11.3 mph


**Question 2 Answer**:

It seems that Alan drove on the highway. This can easily be seen by just observing the data sets, Alan was cruising above 40 mph for a little over three minutes, peaking at 65 mph, then going to 0, perhaps exiting the freeway.  Barbara on the other hand peaked at around 30 mph, a speed that would get you pulled over if being driven on the highway. She also had multiple periods of 0 mph for 30 - 90 seconds at a time, probably indicating she was at a stop light.

Mathematically, Alan averaged a higher speed than Barbara, 32.6 versus 11.3 mph for each trip respectively also suggesting Alan was on the highway for his trip.